File: lib/src/test/java/org/team100/lib/util/DriveUtilTest.java
package org.team100.lib.util;

import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.Optional;
import java.util.Random;

import org.junit.jupiter.api.Test;
import org.team100.lib.geometry.GeometryUtil;
import org.team100.lib.hid.DriverControl;
import org.team100.lib.motion.drivetrain.kinodynamics.SwerveDriveKinematics100;
import org.team100.lib.motion.drivetrain.kinodynamics.SwerveModuleDelta;
import org.team100.lib.motion.drivetrain.kinodynamics.SwerveModuleDeltas;
import org.team100.lib.motion.drivetrain.kinodynamics.SwerveModulePosition100;

import edu.wpi.first.math.geometry.Translation2d;
import edu.wpi.first.math.geometry.Twist2d;

class DriveUtilTest {
    private static final double kDelta = 0.001;

    @Test
    void testClampTwist() {
        {
            // zero is no-op
            DriverControl.Velocity input = new DriverControl.Velocity(0, 0, 0);
            DriverControl.Velocity actual = DriveUtil.clampTwist(input, 1);
            assertEquals(0, actual.x(), kDelta);
            assertEquals(0, actual.y(), kDelta);
            assertEquals(0, actual.theta(), kDelta);
        }
        {
            // clip to the unit circle.
            DriverControl.Velocity input = new DriverControl.Velocity(1, 1, 0);
            DriverControl.Velocity actual = DriveUtil.clampTwist(input, 1);
            assertEquals(0.707, actual.x(), kDelta);
            assertEquals(0.707, actual.y(), kDelta);
        }
        {
            // leave the inside alone
            DriverControl.Velocity input = new DriverControl.Velocity(0.5, 0.5, 0);
            DriverControl.Velocity actual = DriveUtil.clampTwist(input, 1);
            assertEquals(0.5, actual.x(), kDelta);
            assertEquals(0.5, actual.y(), kDelta);
        }
    }

    @Test
    void testRoundTripModuleDeltas() {
        SwerveDriveKinematics100 m_kinematics = new SwerveDriveKinematics100(
                new Translation2d(0.5, 0.5),
                new Translation2d(0.5, -0.5),
                new Translation2d(-0.5, 0.5),
                new Translation2d(-0.5, -0.5));

        {
            // straight diagonal path
            Twist2d t = new Twist2d(1, 1, 0);
            SwerveModuleDeltas p = m_kinematics.toSwerveModuleDelta(t);
            Twist2d t2 = m_kinematics.toTwist2d(p);
            assertEquals(t, t2);
        }
        {
            // turning and moving
            Twist2d t = new Twist2d(1, 1, 1);
            SwerveModuleDeltas p = m_kinematics.toSwerveModuleDelta(t);
            Twist2d t2 = m_kinematics.toTwist2d(p);
            assertEquals(t, t2);
        }
        Random random = new Random(0);
        for (int i = 0; i < 500; ++i) {
            // inverse always works
            Twist2d t = new Twist2d(
                    random.nextDouble(),
                    random.nextDouble(),
                    random.nextDouble());
            SwerveModuleDeltas p = m_kinematics.toSwerveModuleDelta(t);
            Twist2d t2 = m_kinematics.toTwist2d(p);
            assertEquals(t, t2);
        }
    }

    @Test
    void testOneModule() {
        {
            SwerveModulePosition100 start = new SwerveModulePosition100(
                    0, Optional.of(GeometryUtil.kRotationZero));
            SwerveModulePosition100 end = new SwerveModulePosition100(
                    0, Optional.of(GeometryUtil.kRotationZero));
            SwerveModuleDelta delta = DriveUtil.delta(start, end);
            assertEquals(0, delta.distanceMeters, kDelta);
            assertEquals(0, delta.angle.get().getRadians(), kDelta);
        }
        {
            SwerveModulePosition100 start = new SwerveModulePosition100(
                    0, Optional.of(GeometryUtil.kRotationZero));
            SwerveModulePosition100 end = new SwerveModulePosition100(
                    1, Optional.of(GeometryUtil.kRotationZero));
            SwerveModuleDelta delta = DriveUtil.delta(start, end);
            assertEquals(1, delta.distanceMeters, kDelta);
            assertEquals(0, delta.angle.get().getRadians(), kDelta);
        }
        {
            // this ignores the initial zero rotation, and acts as if
            // the path is at 90 degrees the whole time.
            SwerveModulePosition100 start = new SwerveModulePosition100(
                    0, Optional.of(GeometryUtil.kRotationZero));
            SwerveModulePosition100 end = new SwerveModulePosition100(
                    1, Optional.of(GeometryUtil.kRotation90));
            SwerveModuleDelta delta = DriveUtil.delta(start, end);
            assertEquals(1, delta.distanceMeters, kDelta);
            assertEquals(1.571, delta.angle.get().getRadians(), kDelta);
        }
    }

}

File: raspberry_pi/tests/util/drive_util_test.py
import unittest

import random

from wpimath.geometry import Rotation2d, Translation2d, Twist2d

from app.util.drive_util import DriveUtil
from app.kinodynamics.swerve_drive_kinematics import SwerveDriveKinematics100
from app.kinodynamics.swerve_module_position import (
    OptionalRotation2d,
    SwerveModulePosition100,
    SwerveModulePositions,
)

# pylint: disable=C0200,R0903


class DriveUtilTest(unittest.TestCase):
    def test_round_trip_module_deltas(self) -> None:
        k = SwerveDriveKinematics100(
            [
                Translation2d(0.5, 0.5),
                Translation2d(0.5, -0.5),
                Translation2d(-0.5, 0.5),
                Translation2d(-0.5, -0.5),
            ]
        )
        self.assertEqual(4, k.num_modules)

        # straight diagonal path
        t = Twist2d(1, 1, 0)
        p = k.to_swerve_module_delta(t)
        t2 = k.to_twist_2d(p)
        self.assertAlmostEqual(t.dx, t2.dx)
        self.assertAlmostEqual(t.dy, t2.dy)
        self.assertAlmostEqual(t.dtheta, t2.dtheta)

        # turning and moving
        t = Twist2d(1, 1, 1)
        p = k.to_swerve_module_delta(t)
        t2 = k.to_twist_2d(p)
        self.assertAlmostEqual(t.dx, t2.dx)
        self.assertAlmostEqual(t.dy, t2.dy)
        self.assertAlmostEqual(t.dtheta, t2.dtheta)

        for _ in range(500):
            # inverse always works
            t = Twist2d(random.random(), random.random(), random.random())
            p = k.to_swerve_module_delta(t)
            t2 = k.to_twist_2d(p)
            self.assertAlmostEqual(t.dx, t2.dx)
            self.assertAlmostEqual(t.dy, t2.dy)
            self.assertAlmostEqual(t.dtheta, t2.dtheta)

    def test_delta(self) -> None:
        start = SwerveModulePositions(
            SwerveModulePosition100(0, OptionalRotation2d(True, Rotation2d(0))),
            SwerveModulePosition100(0, OptionalRotation2d(True, Rotation2d(0))),
            SwerveModulePosition100(0, OptionalRotation2d(True, Rotation2d(0))),
            SwerveModulePosition100(0, OptionalRotation2d(True, Rotation2d(0))),
        )
        end = SwerveModulePositions(
            SwerveModulePosition100(0, OptionalRotation2d(True, Rotation2d(0))),
            SwerveModulePosition100(0, OptionalRotation2d(True, Rotation2d(0))),
            SwerveModulePosition100(0, OptionalRotation2d(True, Rotation2d(0))),
            SwerveModulePosition100(0, OptionalRotation2d(True, Rotation2d(0))),
        )
        delta = DriveUtil.module_position_delta(start, end)
        self.assertAlmostEqual(0, delta.front_left.distance_m)
        self.assertAlmostEqual(0, delta.front_left.angle.value.radians())

    def test_delta2(self) -> None:
        start = SwerveModulePositions(
            SwerveModulePosition100(0, OptionalRotation2d(True, Rotation2d(0))),
            SwerveModulePosition100(0, OptionalRotation2d(True, Rotation2d(0))),
            SwerveModulePosition100(0, OptionalRotation2d(True, Rotation2d(0))),
            SwerveModulePosition100(0, OptionalRotation2d(True, Rotation2d(0))),
        )
        end = SwerveModulePositions(
            SwerveModulePosition100(0, OptionalRotation2d(True, Rotation2d(1))),
            SwerveModulePosition100(0, OptionalRotation2d(True, Rotation2d(1))),
            SwerveModulePosition100(0, OptionalRotation2d(True, Rotation2d(1))),
            SwerveModulePosition100(0, OptionalRotation2d(True, Rotation2d(1))),
        )
        # there is no interpolation, we use the ending value.
        delta = DriveUtil.module_position_delta(start, end)
        self.assertAlmostEqual(0, delta.front_left.distance_m)
        self.assertAlmostEqual(1, delta.front_left.angle.value.radians())

    def test_delta3(self) -> None:
        start = SwerveModulePositions(
            SwerveModulePosition100(0, OptionalRotation2d(True, Rotation2d(0))),
            SwerveModulePosition100(0, OptionalRotation2d(True, Rotation2d(0))),
            SwerveModulePosition100(0, OptionalRotation2d(True, Rotation2d(0))),
            SwerveModulePosition100(0, OptionalRotation2d(True, Rotation2d(0))),
        )
        end = SwerveModulePositions(
            SwerveModulePosition100(1, OptionalRotation2d(True, Rotation2d(1))),
            SwerveModulePosition100(1, OptionalRotation2d(True, Rotation2d(1))),
            SwerveModulePosition100(1, OptionalRotation2d(True, Rotation2d(1))),
            SwerveModulePosition100(1, OptionalRotation2d(True, Rotation2d(1))),
        )
        # there is no interpolation, we use the ending value.
        delta = DriveUtil.module_position_delta(start, end)
        self.assertAlmostEqual(1, delta.front_left.distance_m)
        self.assertAlmostEqual(1, delta.front_left.angle.value.radians())
